### PR TITLE
feat: further work on dynamic router

### PR DESCRIPTION
**Pull request by jon-nfc**
### :link: Links / References

- Related: #227 

### :construction_worker: Tasks

 - [ ] Add your tasks here as required (delete task if n/a)

- [ ] :orange_circle: Related issue(s) closed via [commit message](https://www.conventionalcommits.org/en/v1.0.0) footer

- [ ] :firecracker: Contains `breaking-change` Any Breaking change(s)? [commit message](https://www.conventionalcommits.org/en/v1.0.0)

    _Breaking Change must also be notated in the commit that introduces it and in [Conventional Commit Format](https://www.conventionalcommits.org/en/v1.0.0/)._

    - [ ] :notebook: Release notes updated

- [ ] :blue_book: Documentation written

    _All features to be documented within the correct section(s). Administration, Development and/or User_

- [ ] :checkered_flag: Milestone assigned

- [ ] :test_tube: [Unit Test(s) Written](https://nofusscomputing.com/projects/centurion_erp/development/testing/)

    _ensure test coverage delta is not less than zero_

- [ ] :page_facing_up: Roadmap updated

### CODE REVIEW

File: src/hooks/useDjangoFetcher.ts [hunk 1]
```diff
@@ -89,17 +108,19 @@ export default async function useDjangoFetcher({
...
+    const backendURL: string = context?.get(backendURLRouteContext)
 
     const options: HTTPNamedParams = {
         credentials: true,
         headers: {
             ...( getCookie( 'csrftoken' ) ? { 'X-CSRFToken': getCookie( 'csrftoken' ) } : {} )
         },
         signal: signal,
-        url: String(`${window.env.API_URL}${URLSanitize(url)}`)
+        url: String(`${(backendURL ? backendURL : window.env.API_URL)}${URLSanitize(url)}`)
```
Comment (jon-nfc) [inline src/hooks/useDjangoFetcher.ts]: once all code refactored to use the newer page loaders under `src/App/pageLoaders`, usage of env var here to be removed so that only context url is used

File: src/App/router.ts [hunk 2]
```diff
@@ -305,28 +465,37 @@ const dynamicRouter = () => {
...
-                                HydrateFallback: () => StateSplash({titleText: "Loading UI", icon: StateIcon.loading }),
+                                id: "notifications",
+                                Component: NotificationLayout,
                                 children: [
                                     {
-                                        id: "page",
-                                        Component: PageContent,
-                                        ErrorBoundary: RouteErrorBoundary,
+                                        id: "UI",
+                                        Component: UI,
```
Comment (jon-nfc) [inline src/App/router.ts]: check if switching ui to use `useRouteLoaderData` sticks to using the route context applicable to this route, and not that of a child route as has been happening with `backend_url`.



- ref: https://reactrouter.com/7.18.2/api/hooks/useRouteLoaderData

File: src/functions/http.ts [hunk 3]
```diff
@@ -78,7 +82,9 @@ export interface HTTPNamedParams {
  */
 export async function httpRequest({
     body = null,
-    credentials = false,
+    credentials = 'omit',
+    referrerPolicy = 'strict-origin-when-cross-origin',
+    mode = 'no-cors',
```
Comment (jon-nfc) [inline src/functions/http.ts]: Ensure that a test exists to check that if none of `credentials`, `referrerPolicy` or `mode` are set that secure defaults are used

File: src/functions/httpJson.ts [hunk 4]
```diff
@@ -38,7 +36,9 @@ from "./http";
  */
 export default async function jsonHttpRequest({
     body = null,
-    credentials = false,
+    credentials = 'omit',
+    referrerPolicy = 'strict-origin-when-cross-origin',
+    mode = 'no-cors',
```
Comment (jon-nfc) [inline src/functions/httpJson.ts]: these don't require defaults outside of `undefined` as the base fetcher is tested and sets the defaults.

add a test to ensure that if none supplied, that undefined is used.

File: src/App/providers/backend.tsx [hunk 5]
```diff
@@ -85,10 +86,31 @@ export interface BackendProviderProps {
...
+
+        if( routes.length > 0 && url === null ) {
+
+            for( let i = (routes.length - 1); i > 0;  i-- ) {
+
+                if( ! Object(routes[i]).hasOwnProperty('handle') ) continue;
+
+                if( ! Object(routes[i].handle).hasOwnProperty('backend_url') ) continue;
+
+                setURL(routes[i].handle.backend_url);
```
Comment (jon-nfc) [inline src/App/providers/backend.tsx:106]: # ToDo

- [ ] Ensure that wherever `backend_url` is specified, that the provider is added to wrap the object.

File: src/hooks/index.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * {@inheritDoc HTTPNamedParams.url}
+     */
+    url: HTTPNamedParams["url"]
+
+    /**
+     * 
+     */
+    baseURL?: HTTPNamedParams["url"]
```
Comment (jon-nfc) [inline src/hooks/index.ts:25]: add inherit doc

File: src/hooks/useDjangoFetcher.ts [hunk 7]
```diff
@@ -85,21 +80,24 @@ export interface djangoFetcherNamedParams {
...
+    baseURL = undefined,
     body = null,
     method = 'GET',
     getMetadata = false,
     onlyMetadata = false,
     signal = null
 }: djangoFetcherNamedParams ): Promise<{ apiData: Response, apiMetadata: Response }> {
 
-
     const options: HTTPNamedParams = {
```
Comment (jon-nfc) [inline src/hooks/useDjangoFetcher.ts:91]: Add logic so that if `baseURL = undefined` that the `BackendProvider` context is used to  obtain the URL.

User should not be able to supply if `BackendProvider` available. (means that it's within the app and not a route, i.e. loader)

File: src/App/providers/backend.tsx (hunk 5 above)
Comment (jon-nfc) [reply src/App/providers/backend.tsx:106]: done c93fb559a90f761ebc623f7deeb3165a4a48a3e8

File: src/layout/Markdown.tsx [hunk 8]
```diff
@@ -300,7 +304,15 @@ const Markdown = (): React.JSX.Element => {
...
-                                    {markdownDocument}
+                                    {String(
+                                        markdownDocument
+                                    ).replaceAll(
+                                        '(./', `(${document.location.pathname}/`
+                                    ).replaceAll(
+                                        'index.md)', ')'
+                                    ).replaceAll(
+                                        '.md)', ')'
+                                    )}
```
Comment (jon-nfc) [inline src/layout/Markdown.tsx:315]: Add tests to ensure that the correct URLs are rendered.

File: src/types/backend/apiMetadata/RouteDescriptions.ts [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+ * @category Type
+ * @since 0.13.0
+ */
+export type RouteActionDescription =
+    
+    /**
+     * Use API FormSubmission.
+     */
+    "api"
+
```
Comment (jon-nfc) [inline src/types/backend/apiMetadata/RouteDescriptions.ts:18]: all types that are "descriptions" need their doc category to be  `@category Description`

File: src/types/window.d.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { UIEnvironment } from "./backend/envBackend";
+
+// eslint-disable-next-line
+export {};
+
+declare global {
+    interface Window {
+        env: UIEnvironment;
+    }
+}
```
Comment (jon-nfc) [inline src/types/window.d.ts]: add nl to end of file

File: src/App/router.ts (hunk 2 above)
Comment (jon-nfc) [reply src/App/router.ts]: no longer relevant.

File: src/hooks/useDjangoFetcher.ts (hunk 1 above)
Comment (jon-nfc) [reply src/hooks/useDjangoFetcher.ts]: addressed in #235

File: src/layout/Markdown.tsx (hunk 8 above)
Comment (jon-nfc) [reply src/layout/Markdown.tsx:315]: added to #207 as feat.

File: src/types/backend/apiMetadata/RouteDescriptions.ts (hunk 9 above)
Comment (jon-nfc) [reply src/types/backend/apiMetadata/RouteDescriptions.ts:18]: already addressed.

File: src/types/window.d.ts (hunk 10 above)
Comment (jon-nfc) [reply src/types/window.d.ts]: 54aea522c11d0c8f280f8045abfa13c3adc87bb3

File: src/hooks/index.ts (hunk 6 above)
Comment (jon-nfc) [reply src/hooks/index.ts:25]: dfa6b7687edc4b31f50e9ea9fe947023add7e303

File: src/hooks/useDjangoFetcher.ts (hunk 7 above)
Comment (jon-nfc) [reply src/hooks/useDjangoFetcher.ts:91]: addressed in #235

File: src/functions/http.ts (hunk 3 above)
Comment (jon-nfc) [reply src/functions/http.ts]: 84011ba9751d3b6ddb8b0c93a5f1af0613bc0982

File: src/functions/httpJson.ts (hunk 4 above)
Comment (jon-nfc) [reply src/functions/httpJson.ts]: 84011ba9751d3b6ddb8b0c93a5f1af0613bc0982
